### DISCUSSION TITLE
Single Board Computer I2C Connection

### DISCUSSION THREAD

**Discussion by pkr97** [fjp/2wd-robot#78] [Q&A]
Hi Franz,
I find it a bit confusing that the MCU is going to handle the Motor encoder and also Actuation. The MCU should be receiving commands and possibly sending sensory readings to the SBC as shown here https://ros-mobile-robots.com/fritzing/remo_architecture.svg

While going through the Hardware Interfaces section, I found that the SBC needs to communicate with the motor driver. This contradicts the figure shown here https://ros-mobile-robots.com/fritzing/remo_architecture.svg 

This brings some ambiguity that how to connect all the hardware for minimal testing at the first place.

Best.....

**Comment by fjp** [comment]
Hi @pkr97, first let me clarify that in the docs are parts that refer to the low-level approach and some parts to the high-level approach. This is not so nice and I hope to get back more time to fix this soon. So without knowing which parts you are referring to in the docs it is a bit hard for me to give proper feedback. Anyway let's try:

> I found that the SBC needs to communicate with the motor driver

Here I assume with you mean the https://github.com/ros-mobile-robots/grove_motor_driver which is the motor driver that is used for the diffbot prototype I built first when I started this project.

![](https://ros-mobile-robots.com/fritzing/diffbot_architecture.svg)

The [`write()` method](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/src/diffbot_hw_interface.cpp#L163) in the hardware interface serves currently two interfaces (which is a bit confusing): 

- [high-level approach](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/src/diffbot_hw_interface.cpp#L188-L227) - [diffbot schematic](https://ros-mobile-robots.com/fritzing/diffbot_architecture.svg)
- [low-level approach](https://github.com/ros-mobile-robots/diffbot/blob/noetic-devel/diffbot_base/src/diffbot_hw_interface.cpp#L174-L186) - [remo schematic](https://ros-mobile-robots.com/fritzing/remo_architecture.svg)

The high-level approach that is based on the schematic shown above and is for the "old" diffbot model.

The [link you posted of the new remo robot schematic](https://ros-mobile-robots.com/fritzing/remo_architecture.svg) is using the low-level approach. I recently explained the difference between the high and low level approaches here: https://github.com/ros-mobile-robots/diffbot/discussions/76#discussioncomment-4975667


> I find it a bit confusing that the MCU is going to handle the Motor encoder and also Actuation.

The reason for not processing the encoder ticks on the SBC is that it might miss some ticks if the rate of ticks is too high. Usually a mcu is a better fit for this becasue its pins should be able to handle higher frequencies. Anyway depending on the encoders this might work as well if the encoders are directly connected to the SBC. For diffbot and remo robot, the encoders ware always connected to the mcu and then those sensor information was read via rosserial on the SBC. 

As mentioned above, for diffbot, I used the approach you mention: that the SBC communicates with the (grove) motor driver and receives directly the motor commands from the ROS control hardware interface. The current approach I use for remo robot is to compute "high-level" control commands using ROS control and the diff_drive_controller (which works with a VelocityInterface) and then send those "high-level" velocity commands to the mcu where it gets converted to motor commands. Which approach to choose is personal preference I would say and both is possible.

In the remo schematic the mcu receives the high level velocity commands from the SBC via USB and then communicates with the motro driver board (FeatherWing) via i2c (yellow and white wires; the blue pcb with the four white connectors).

**Reply by fjp** [reply]
> This brings some ambiguity that how to connect all the hardware for minimal testing at the first place.

In the https://github.com/ros-mobile-robots/diffbot/tree/noetic-devel/diffbot_base/scripts folder you find some sub folders named `test`, these should provide some good starting points to flash an mcu and test for example only the encoders, or the motor driver board.

**Reply by fjp** [reply]
I hope this clarifies some of your confusion, otherwise let me know which part I might missed to explain better.

**Reply by pkr97** [reply]
Hi Franz,
thanks a ton for the detailed reply.

I was trying to build a REMO and have been failing because of the software packages hosted in this main repository.

Although your answer is very much understandable, it still will create confusion for a newbie in handling the ROS packages.

Maybe you could have an instruction sheet for getting started. That would help a lot and as you will be well aware of problems while setting on your own, it would help in framing the instruction sheet. And anytime the additional content can be pointed by using a link to the existing page that hosts a huge resource for building and running the robot.

This detailed instruction sheet would help many and me also in building a REMO successfully.

Cheers.

**Reply by fjp** [reply]
Hi @pkr97, thanks a lot for the feedback and the idea with the instruction sheet. My plan is to structure the documentation a bit more clearer and put some getting started pages at the beginning, that can act as a instruction sheet. I will also think of providing this as a pdf which should make it more straigth forward to follow.

Additionally I've some raw footage of instructions, showing the terminal commands and the robot actions, but this needs time to edit and put in the right place. With some patience, I hope these additions will make this project more useful and easier to understand.

For now, always feel free to start a discussion or open an issue to clarify things. What also helps to improve the project, is if you can point to specific sections in the documentation that need clarification, or even better: you can open a PR yourself in the docs repo: https://github.com/ros-mobile-robots/ros-mobile-robots.github.io. This can also easily be done from every page of the docs, using the Edit button:

![image](https://user-images.githubusercontent.com/1286618/221373752-9e20fe21-ea2f-411b-a57c-eddb31d73e3d.png)
 
Best regards
Franz
